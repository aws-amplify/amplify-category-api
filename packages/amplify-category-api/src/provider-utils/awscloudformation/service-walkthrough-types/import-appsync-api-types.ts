export enum ImportedRDSType {
  MYSQL = 'mysql',
  POSTGRESQL= 'postgresql'
};

export type ImportedDataSourceType = ImportedRDSType;

export type ImportAppSyncAPIInputs =  {
  apiName: string,
  dataSourceType: ImportedDataSourceType
};

export const RDS_SCHEMA_FILE_NAME = 'schema.rds.graphql';
