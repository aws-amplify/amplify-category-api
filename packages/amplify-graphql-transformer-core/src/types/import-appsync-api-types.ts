import { TransformerSecrets } from '@aws-amplify/graphql-transformer-interfaces';

export enum ImportedRDSType {
  MYSQL = 'mysql',
  POSTGRESQL= 'postgresql'
};

export type ImportedDataSourceType = ImportedRDSType;

export type ImportedDataSourceConfig = RDSDataSourceConfig;
export type RDSDataSourceConfig = RDSConnectionSecrets & {
  engine: ImportedRDSType
};

export type ImportAppSyncAPIInputs = {
  apiName: string,
  dataSourceConfig: ImportedDataSourceConfig
};

export const RDS_SCHEMA_FILE_NAME = 'schema.rds.graphql';

export type RDSConnectionSecrets = TransformerSecrets & {
  username: string,
  password: string,
  host: string,
  database: string,
  port: number,
};

export const MYSQL_DB_TYPE = 'MySQL';
export const DDB_DB_TYPE = 'DDB';
