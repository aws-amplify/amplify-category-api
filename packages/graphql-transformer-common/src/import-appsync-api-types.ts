// Types in this file represent the values discovered during the CLI `amplify import api` flow

/**
 * The database engine type as recorded in the generated GraphQL schema. This type is only used for the walkthrough and schema generation
 * flows. The transformer internals use the ModelDataSourceDbType.
 *
 * This enum should not be exported to the category API.
 */
export enum ImportedRDSType {
  MYSQL = 'mysql',
  POSTGRESQL = 'postgres',
}

export type ImportedDataSourceConfig = {
  database: string;
  engine: ImportedRDSType;
  host: string;
  password: string;
  port: number;
  username: string;
};

export type ImportAppSyncAPIInputs = {
  apiName: string;
  dataSourceConfig?: ImportedDataSourceConfig;
};
