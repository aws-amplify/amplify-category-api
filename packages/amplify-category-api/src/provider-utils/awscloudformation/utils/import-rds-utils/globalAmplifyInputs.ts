import { ImportedRDSType } from '../../service-walkthrough-types/import-appsync-api-types';
import { parse } from 'graphql';
import * as fs from 'fs-extra';
import { $TSAny, $TSContext, ApiCategoryFacade, getGraphQLTransformerAuthDocLink } from 'amplify-cli-core';
import _ from 'lodash';

type AmplifyInputEntry = {
  name: string,
  type: string,
  default: string|number,
  comment?: string|undefined
};

const getGlobalAmplifyInputEntries = async (
  context: $TSContext, 
  dataSourceType = ImportedRDSType.MYSQL, 
  includeAuthRule = true
  ): Promise<AmplifyInputEntry[]> => {
  const inputs: AmplifyInputEntry[] = [
    { 
      name: 'engine',
      type: 'String',
      default: dataSourceType,
    },
    {
      name: 'host',
      type: 'String',
      default: 'ENTER YOUR DATABASE HOSTNAME HERE',
    },
    {
      name: 'port',
      type: 'Int',
      default: 3306,
      comment: 'ENTER PORT NUMBER HERE'
    },
    {
      name: 'database',
      type: 'String',
      default: 'ENTER YOUR DATABASE NAME HERE'
    }
  ];

  if (includeAuthRule && (await ApiCategoryFacade.getTransformerVersion(context) === 2)) {
    const authDocLink = getGraphQLTransformerAuthDocLink(2);
    inputs.push({
      name: 'globalAuthRule',
      type: 'AuthRule',
      default: '{ allow: public }',
      comment: `This "input" configures a global authorization rule to enable public access to all models in this schema. Learn more about authorization rules here:${authDocLink}`
    });
  };
  return inputs;
};

export const constructGlobalAmplifyInput = async (context: $TSContext, dataSourceType: ImportedRDSType) => {
  const inputs = await getGlobalAmplifyInputEntries(context, dataSourceType);
  const inputsString = inputs.reduce((acc: string, input): string =>
    acc + ` ${input.name}: ${input.type} = ${input.type === 'String' ? '"'+ input.default + '"' : input.default} ${input.comment ? '# ' + input.comment: ''} \n`
  , '');
  return `input Amplify {\n${inputsString}}\n`;
};

export const readGlobalAmplifyInput = async (context: $TSContext, pathToSchemaFile: string) => {
  const schemaContent = fs.readFileSync(pathToSchemaFile, 'utf-8');
  if(_.isEmpty(schemaContent?.replace(/[\r\n]/gm, ''))) {
    throw new Error('The schema file is empty');
  }

  const parsedSchema = parse(schemaContent);

  const inputDirective: $TSAny = parsedSchema.definitions.find(
    definition =>
     definition.kind === 'InputObjectTypeDefinition' &&
     definition.name &&
     definition.name.value === 'Amplify'
  );

  const expectedInputs = (await getGlobalAmplifyInputEntries(context, ImportedRDSType.MYSQL, false)).map(item => item.name);
  const inputs = {};
  expectedInputs.map(input => {
    const value = inputDirective?.fields?.find( 
      field => field?.name?.value === input
    ).defaultValue?.value;
    if (_.isEmpty(value)) {
      throw new Error(`Invalid value for ${input} input in the GraphQL schema. Correct and re-try.`);
    }
    inputs[input] = value;
  });

  return inputs;
};

export const validateInputConfig = async (context: $TSContext, config: { [x: string]: any; }) => {
  const expectedInputs = (await getGlobalAmplifyInputEntries(context, ImportedRDSType.MYSQL, false)).map(item => item.name);
  const missingInputs = expectedInputs.filter(input => _.isEmpty(config[input]));
  if(!_.isEmpty(missingInputs)) {
    throw new Error(`The input parameters ${missingInputs.join(',')} are missing. Specify them in the schema and re-run.`);
  }

  // The database connection details shouldn't have space
  const invalidInputs = expectedInputs.filter(input => config[input]?.indexOf(' ') >= 0);
  if(!_.isEmpty(invalidInputs)) {
    throw new Error(`The database connection parameters ${invalidInputs.join(',')} might be invalid. Correct them in the schema and re-run.`);
  }
};
