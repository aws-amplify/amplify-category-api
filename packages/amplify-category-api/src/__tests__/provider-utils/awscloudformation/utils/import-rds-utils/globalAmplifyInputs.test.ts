import { $TSAny, $TSContext } from 'amplify-cli-core';
import { constructDefaultGlobalAmplifyInput, getRDSGlobalAmplifyInput, getRDSDBConfigFromAmplifyInput, validateRDSInputDBConfig, readRDSGlobalAmplifyInput, constructRDSGlobalAmplifyInput } from '../../../../../provider-utils/awscloudformation/utils/import-rds-utils/globalAmplifyInputs';
import { ImportedRDSType } from '../../../../../provider-utils/awscloudformation/service-walkthrough-types/import-appsync-api-types';
import * as fs from 'fs-extra';

jest.mock('fs-extra', () => ({
  readFileSync: jest.fn(),
}));
const readFileSync_mock = fs.readFileSync as jest.MockedFunction<typeof fs.readFileSync>;

jest.mock('amplify-cli-core', () => {
  const original = jest.requireActual('amplify-cli-core');
  return {
    ...original,
    ApiCategoryFacade: {
      getTransformerVersion: jest.fn().mockResolvedValue(2),
    },
  };
});

describe('Amplify Input read/write from schema', () => {
  const mockContext = {} as $TSAny as $TSContext;

  afterAll(() => {
    jest.resetAllMocks();
  });

  it('constructs valid default input parameters for MySQL datasource with global auth rule', async () => {
    const expectedGraphQLInputString = `input Amplify {
      engine: String = \"mysql\"  
      host: String = \"ENTER YOUR DATABASE HOSTNAME HERE\"  
      port: Int = 3306 # ENTER PORT NUMBER HERE
      database: String = \"ENTER YOUR DATABASE NAME HERE\" 
      globalAuthRule: AuthRule = { allow: public } # This "input" configures a global authorization rule to enable public access to all models in this schema. Learn more about authorization rules here:https://docs.amplify.aws/cli/graphql/authorization-rules 
    }`;
    const constructedInputString = await constructDefaultGlobalAmplifyInput(mockContext, ImportedRDSType.MYSQL);
    expect(constructedInputString?.replace(/\s/g, '')).toEqual(expectedGraphQLInputString.replace(/\s/g, ''));
  });

  it('reads global Amplify input type in the schema', async () => {
    const mockValidInputs = {
      engine: 'mysql',
      host: 'mockdatabase.rds.amazonaws.com',
      port: '1010',
      database: 'mockdatabase'
    };

    const mockInputSchema = `input Amplify {
      engine: String = \"${mockValidInputs.engine}\"  
      host: String = \"${mockValidInputs.host}\"  
      port: Int = ${mockValidInputs.port} # ENTER PORT NUMBER HERE
      database: String = \"${mockValidInputs.database}\" 
      globalAuthRule: AuthRule = { allow: public } # This "input" configures a global authorization rule to enable public access to all models in this schema. Learn more about authorization rules here:https://docs.amplify.aws/cli/graphql/authorization-rules 
    }`;

    readFileSync_mock.mockReturnValue(mockInputSchema);

    const readInputNode = await readRDSGlobalAmplifyInput('mock/path');
    expect(readInputNode).toMatchSnapshot();
  });

  it('fetch DB config from the input arguments for database connection details', async () => {
    const mockValidInputs = {
      engine: 'mysql',
      host: 'mockdatabase.rds.amazonaws.com',
      port: '1010',
      database: 'mockdatabase'
    };

    const mockInputSchema = `input Amplify {
      engine: String = \"${mockValidInputs.engine}\"  
      host: String = \"${mockValidInputs.host}\"  
      port: Int = ${mockValidInputs.port} # ENTER PORT NUMBER HERE
      database: String = \"${mockValidInputs.database}\" 
      globalAuthRule: AuthRule = { allow: public } # This "input" configures a global authorization rule to enable public access to all models in this schema. Learn more about authorization rules here:https://docs.amplify.aws/cli/graphql/authorization-rules 
    }`;
    readFileSync_mock.mockReturnValue(mockInputSchema);

    const readInputs = await getRDSGlobalAmplifyInput(mockContext, '');
    const readConfig = await getRDSDBConfigFromAmplifyInput(mockContext, readInputs);
    
    expect(readConfig).toEqual(mockValidInputs);
  });

  it('constructs the global Amplify input from given config', async () => {
    const mockValidInputs = {
      engine: 'mysql',
      host: 'mockdatabase.rds.amazonaws.com',
      port: '1010',
      database: 'mockdatabase'
    };

    const mockInputSchema = `input Amplify {
      engine: String = \"${mockValidInputs.engine}\"  
      host: String = \"${mockValidInputs.host}\"  
      port: Int = ${mockValidInputs.port} # ENTER PORT NUMBER HERE
      database: String = \"${mockValidInputs.database}\" 
      globalAuthRule: AuthRule = { allow: public } # This "input" configures a global authorization rule to enable public access to all models in this schema. Learn more about authorization rules here:https://docs.amplify.aws/cli/graphql/authorization-rules 
    }`;

    readFileSync_mock.mockReturnValue(mockInputSchema);

    const readInputNode = await readRDSGlobalAmplifyInput('mock/path');
    
    const userInputs = {
      host: 'otherdatabase.rds.amazonaws.com',
      port: 2020,
      database: 'otherdatabase'
    };

    const constructedInputDefinition = await constructRDSGlobalAmplifyInput(mockContext, userInputs, readInputNode);
    expect(constructedInputDefinition).toMatchSnapshot();
  });

  it('reports missing input arguments for database connection details', async () => {
    const mockInvalidInputs = {
      engine: 'mysql',
      port: '1010',
      database: 'mockdatabase'
    };

    try {
      await validateRDSInputDBConfig(mockContext, mockInvalidInputs);
      fail('invalid input configuration is not reported');
    }
    catch(error) {
      expect(error?.message).toEqual('The database connection parameters: host, are missing. Specify them in the schema and re-run.');
    }
  });

  it('reports invalid input arguments for database connection details', async () => {
    const mockInvalidInputs = {
      engine: 'mysql',
      host: 'ENTER YOUR DATABASE HOSTNAME HERE', //invalid entry
      port: '1010',
      database: 'mockdatabase'
    };

    try {
      await validateRDSInputDBConfig(mockContext, mockInvalidInputs);
      fail('invalid input configuration is not reported');
    }
    catch(error) {
      expect(error?.message).toEqual('The database connection parameters: host, might be invalid. Correct them in the schema and re-run.');
    }
  });
});
