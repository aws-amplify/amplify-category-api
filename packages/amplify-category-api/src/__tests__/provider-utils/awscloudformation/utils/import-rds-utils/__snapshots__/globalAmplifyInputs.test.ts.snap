// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amplify Input read/write from schema constructs the global Amplify input from given config 1`] = `
"input Amplify {
  engine: String = \\"mysql\\"
  host: String = \\"otherdatabase.rds.amazonaws.com\\"
  port: Int = 2020
  database: String = \\"otherdatabase\\"
  globalAuthRule: AuthRule = {allow: public}
}"
`;

exports[`Amplify Input read/write from schema reads global Amplify input type in the schema 1`] = `
Object {
  "description": undefined,
  "directives": Array [],
  "fields": Array [
    Object {
      "defaultValue": Object {
        "block": false,
        "kind": "StringValue",
        "loc": Object {
          "end": 46,
          "start": 39,
        },
        "value": "mysql",
      },
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "loc": Object {
        "end": 46,
        "start": 22,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 28,
          "start": 22,
        },
        "value": "engine",
      },
      "type": Object {
        "kind": "NamedType",
        "loc": Object {
          "end": 36,
          "start": 30,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 36,
            "start": 30,
          },
          "value": "String",
        },
      },
    },
    Object {
      "defaultValue": Object {
        "block": false,
        "kind": "StringValue",
        "loc": Object {
          "end": 102,
          "start": 70,
        },
        "value": "mockdatabase.rds.amazonaws.com",
      },
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "loc": Object {
        "end": 102,
        "start": 55,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 59,
          "start": 55,
        },
        "value": "host",
      },
      "type": Object {
        "kind": "NamedType",
        "loc": Object {
          "end": 67,
          "start": 61,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 67,
            "start": 61,
          },
          "value": "String",
        },
      },
    },
    Object {
      "defaultValue": Object {
        "kind": "IntValue",
        "loc": Object {
          "end": 127,
          "start": 123,
        },
        "value": "1010",
      },
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "loc": Object {
        "end": 127,
        "start": 111,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 115,
          "start": 111,
        },
        "value": "port",
      },
      "type": Object {
        "kind": "NamedType",
        "loc": Object {
          "end": 120,
          "start": 117,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 120,
            "start": 117,
          },
          "value": "Int",
        },
      },
    },
    Object {
      "defaultValue": Object {
        "block": false,
        "kind": "StringValue",
        "loc": Object {
          "end": 192,
          "start": 178,
        },
        "value": "mockdatabase",
      },
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "loc": Object {
        "end": 192,
        "start": 159,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 167,
          "start": 159,
        },
        "value": "database",
      },
      "type": Object {
        "kind": "NamedType",
        "loc": Object {
          "end": 175,
          "start": 169,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 175,
            "start": 169,
          },
          "value": "String",
        },
      },
    },
    Object {
      "defaultValue": Object {
        "fields": Array [
          Object {
            "kind": "ObjectField",
            "loc": Object {
              "end": 242,
              "start": 229,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 234,
                "start": 229,
              },
              "value": "allow",
            },
            "value": Object {
              "kind": "EnumValue",
              "loc": Object {
                "end": 242,
                "start": 236,
              },
              "value": "public",
            },
          },
        ],
        "kind": "ObjectValue",
        "loc": Object {
          "end": 244,
          "start": 227,
        },
      },
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "loc": Object {
        "end": 244,
        "start": 200,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 214,
          "start": 200,
        },
        "value": "globalAuthRule",
      },
      "type": Object {
        "kind": "NamedType",
        "loc": Object {
          "end": 224,
          "start": 216,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 224,
            "start": 216,
          },
          "value": "AuthRule",
        },
      },
    },
  ],
  "kind": "InputObjectTypeDefinition",
  "loc": Object {
    "end": 458,
    "start": 0,
  },
  "name": Object {
    "kind": "Name",
    "loc": Object {
      "end": 13,
      "start": 6,
    },
    "value": "Amplify",
  },
}
`;
